servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/LingeringRoundRobinLoadBalancerTest.java

```diff
@@ -0,0 +1,202 @@
+
+        sendServiceDiscoveryEvents(downEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-1");
+
+        // We validate the host is expired by attempting to create a new connection
+        final Predicate<TestLoadBalancedConnection> createNewConnection = alwaysNewConnectionFilter();
+        Exception e = assertThrows(ExecutionException.class, () ->
+                lb.selectConnection(createNewConnection).toFuture().get());
+        assertThat(e.getCause(), instanceOf(NoAvailableHostException.class));
+
+        lb.selectConnection(any()).toFuture().get();
+
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-1");
+        lb.selectConnection(createNewConnection).toFuture().get();
+
+        assertConnectionCount(lb.activeAddresses(), connectionsCount("address-1", 2));
+    }
+
+    @Test
+    public void handleDiscoveryEventsForConnectedHosts() throws Exception {
+        assertAddresses(lb.activeAddresses(), EMPTY_ARRAY);
+        assertThat(lb.activeAddresses(), is(empty()));
+
+        final Predicate<TestLoadBalancedConnection> connectionFilter = alwaysNewConnectionFilter();
+
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-1");
+        // For an added host, connection needs to be initiated, otherwise the host is free to be deleted
+        lb.selectConnection(connectionFilter).toFuture().get();
+
+        sendServiceDiscoveryEvents(downEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-1");
+
+        sendServiceDiscoveryEvents(upEvent("address-2"));
+        lb.selectConnection(connectionFilter).toFuture().get();
+
+        assertAddresses(lb.activeAddresses(), "address-1", "address-2");
+
+        sendServiceDiscoveryEvents(downEvent("address-3"));
+        assertAddresses(lb.activeAddresses(), "address-1", "address-2");
+
+        // Marking the first host as not expired should not create duplicates
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-1", "address-2");
+
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        // At this moment in time, duplicates are allowed and the down event will remove just the first address
+        assertAddresses(lb.activeAddresses(), "address-1", "address-1", "address-2");
+
+        sendServiceDiscoveryEvents(downEvent("address-1"));
+        // Because the first address (new ones are added at the end) has an open connection,
+        // both addresses stay in the collection
+        assertAddresses(lb.activeAddresses(), "address-1", "address-1", "address-2");
+
+        sendServiceDiscoveryEvents(downEvent("address-2"));
+        // This host has a connection open, so it stays.
+        assertAddresses(lb.activeAddresses(), "address-1", "address-1", "address-2");
```

Comment:
I've played with the tests. Please, make sure that all tests pass after the following changes:

```patch
Index: servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/EagerRoundRobinLoadBalancerTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/EagerRoundRobinLoadBalancerTest.java b/servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/EagerRoundRobinLoadBalancerTest.java
--- a/servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/EagerRoundRobinLoadBalancerTest.java	(revision a486e819a22c225237794acc1cb81be1db03cb95)
+++ b/servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/EagerRoundRobinLoadBalancerTest.java	(date 1626325084267)
@@ -58,7 +58,7 @@
         assertAddresses(lb.activeAddresses(), "address-2");
 
         sendServiceDiscoveryEvents(upEvent("address-1"));
-        assertAddresses(lb.activeAddresses(), "address-1", "address-2");
+        assertAddresses(lb.activeAddresses(), "address-2", "address-1");
 
         sendServiceDiscoveryEvents(downEvent("address-1"));
         assertAddresses(lb.activeAddresses(), "address-2");
Index: servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/LingeringRoundRobinLoadBalancerTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/LingeringRoundRobinLoadBalancerTest.java b/servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/LingeringRoundRobinLoadBalancerTest.java
--- a/servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/LingeringRoundRobinLoadBalancerTest.java	(revision a486e819a22c225237794acc1cb81be1db03cb95)
+++ b/servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/LingeringRoundRobinLoadBalancerTest.java	(date 1626331141590)
@@ -16,10 +16,13 @@
 package io.servicetalk.loadbalancer;
 
 import io.servicetalk.client.api.NoAvailableHostException;
+import io.servicetalk.concurrent.api.Executor;
+import io.servicetalk.concurrent.api.Executors;
 
 import org.junit.Test;
 
 import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
 import java.util.function.Predicate;
 
 import static io.servicetalk.concurrent.internal.DeliberateException.DELIBERATE_EXCEPTION;
@@ -27,6 +30,7 @@
 import static org.hamcrest.Matchers.empty;
 import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
 import static org.junit.Assert.assertThrows;
 import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.times;
@@ -35,7 +39,16 @@
 public class LingeringRoundRobinLoadBalancerTest extends RoundRobinLoadBalancerTest {
 
     @Test
-    public void hostDownDoesntCloseConnection() throws Exception {
+    public void hostDownDoesntCloseConnectionCloseLB() throws Exception {
+        hostDownDoesntCloseConnection(true);
+    }
+
+    @Test
+    public void hostDownDoesntCloseConnectionCloseLBGracefully() throws Exception {
+        hostDownDoesntCloseConnection(true);
+    }
+
+    private void hostDownDoesntCloseConnection(boolean gracefulClosure) throws Exception {
         sendServiceDiscoveryEvents(upEvent("address-1"));
         TestLoadBalancedConnection conn = lb.selectConnection(any()).toFuture().get();
         sendServiceDiscoveryEvents(downEvent("address-1"));
@@ -43,8 +56,15 @@
         verify(conn, never()).closeAsync();
 
         // But connection is cleaned when LB is closed
-        lb.closeAsync().toFuture().get();
-        verify(conn, times(1)).closeAsync();
+        if (gracefulClosure) {
+            lb.closeAsyncGracefully().toFuture().get();
+            verify(conn, times(1)).closeAsyncGracefully();
+            verify(conn, never()).closeAsync();
+        } else {
+            lb.closeAsync().toFuture().get();
+            verify(conn, times(1)).closeAsync();
+            verify(conn, never()).closeAsyncGracefully();
+        }
         assertAddresses(lb.activeAddresses(), EMPTY_ARRAY);
     }
 
@@ -67,6 +87,28 @@
         assertAddresses(lb.activeAddresses(), EMPTY_ARRAY);
     }
 
+    @Test
+    public void closureOfLastConnectionDoesntRaceWithNewAvailableEvent() throws Exception {
+        Executor executor = Executors.newFixedSizeExecutor(1);
+        try {
+            sendServiceDiscoveryEvents(upEvent("address-1"));
+            TestLoadBalancedConnection conn = lb.selectConnection(alwaysNewConnectionFilter()).toFuture().get();
+            sendServiceDiscoveryEvents(downEvent("address-1"));
+            assertConnectionCount(lb.activeAddresses(), connectionsCount("address-1", 1));
+
+            Future<Object> f = executor.submit(() -> {
+                sendServiceDiscoveryEvents(upEvent("address-1"));
+                return null;
+            }).toFuture();
+
+            conn.closeAsync().toFuture().get();
+            f.get();
+            assertConnectionCount(lb.activeAddresses(), connectionsCount("address-1", 0));
+        } finally {
+            executor.closeAsync().toFuture().get();
+        }
+    }
+
     @Test
     public void closedConnectionDoesntRemoveActiveHost() throws Exception {
         sendServiceDiscoveryEvents(upEvent("address-1"));
@@ -89,11 +131,11 @@
         assertAddresses(lb.activeAddresses(), "address-1");
 
         lb.selectConnection(any()).toFuture().get();
-
         assertConnectionCount(lb.activeAddresses(), connectionsCount("address-1", 1));
 
         sendServiceDiscoveryEvents(downEvent("address-1"));
-        assertAddresses(lb.activeAddresses(), "address-1");
+        assertConnectionCount(lb.activeAddresses(), connectionsCount("address-1", 1));
+        assertThat(lb.selectConnection(any()).toFuture().get(), is(notNullValue()));
 
         // We validate the host is expired by attempting to create a new connection
         final Predicate<TestLoadBalancedConnection> createNewConnection = alwaysNewConnectionFilter();
@@ -101,18 +143,15 @@
                 lb.selectConnection(createNewConnection).toFuture().get());
         assertThat(e.getCause(), instanceOf(NoAvailableHostException.class));
 
-        lb.selectConnection(any()).toFuture().get();
-
+        // When the host becomes available again, new connections can be created
         sendServiceDiscoveryEvents(upEvent("address-1"));
         assertAddresses(lb.activeAddresses(), "address-1");
         lb.selectConnection(createNewConnection).toFuture().get();
-
         assertConnectionCount(lb.activeAddresses(), connectionsCount("address-1", 2));
     }
 
     @Test
     public void handleDiscoveryEventsForConnectedHosts() throws Exception {
-        assertAddresses(lb.activeAddresses(), EMPTY_ARRAY);
         assertThat(lb.activeAddresses(), is(empty()));
 
         final Predicate<TestLoadBalancedConnection> connectionFilter = alwaysNewConnectionFilter();
@@ -137,27 +176,30 @@
         sendServiceDiscoveryEvents(upEvent("address-1"));
         assertAddresses(lb.activeAddresses(), "address-1", "address-2");
 
+        // When all hosts are active, new event creates a duplicate
         sendServiceDiscoveryEvents(upEvent("address-1"));
-        // At this moment in time, duplicates are allowed and the down event will remove just the first address
-        assertAddresses(lb.activeAddresses(), "address-1", "address-1", "address-2");
+        assertAddresses(lb.activeAddresses(), "address-1", "address-2", "address-1");
 
-        sendServiceDiscoveryEvents(downEvent("address-1"));
-        // Because the first address (new ones are added at the end) has an open connection,
+        // Because the first address (new ones are added at the end) has an open connection it's marked as "expired",
         // both addresses stay in the collection
-        assertAddresses(lb.activeAddresses(), "address-1", "address-1", "address-2");
+        sendServiceDiscoveryEvents(downEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-1", "address-2", "address-1");
 
+        // This host has a connection open, so it stays as "expired".
         sendServiceDiscoveryEvents(downEvent("address-2"));
-        // This host has a connection open, so it stays.
-        assertAddresses(lb.activeAddresses(), "address-1", "address-1", "address-2");
+        assertAddresses(lb.activeAddresses(), "address-1", "address-2", "address-1");
+
+        // Another "down" event removes the second entry of "address-1" that does not have active connections
+        sendServiceDiscoveryEvents(downEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-1", "address-2");
 
         // Let's make sure that an SD failure doesn't compromise LB's internal state
         serviceDiscoveryPublisher.onError(DELIBERATE_EXCEPTION);
-        assertAddresses(lb.activeAddresses(), "address-1", "address-1", "address-2");
+        assertAddresses(lb.activeAddresses(), "address-1", "address-2", "address-1");
     }
 
     @Test
     public void handleDiscoveryEventsForNotConnectedHosts() {
-        assertAddresses(lb.activeAddresses(), EMPTY_ARRAY);
         assertThat(lb.activeAddresses(), is(empty()));
 
         sendServiceDiscoveryEvents(upEvent("address-1"));
@@ -173,7 +215,7 @@
         assertAddresses(lb.activeAddresses(), "address-2");
 
         sendServiceDiscoveryEvents(upEvent("address-1"));
-        assertAddresses(lb.activeAddresses(), "address-1", "address-2");
+        assertAddresses(lb.activeAddresses(), "address-2", "address-1");
 
         sendServiceDiscoveryEvents(downEvent("address-1"));
         assertAddresses(lb.activeAddresses(), "address-2");
Index: servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerTest.java b/servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerTest.java
--- a/servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerTest.java	(revision a486e819a22c225237794acc1cb81be1db03cb95)
+++ b/servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerTest.java	(date 1626325104652)
@@ -79,7 +79,6 @@
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.both;
 import static org.hamcrest.Matchers.contains;
-import static org.hamcrest.Matchers.containsInAnyOrder;
 import static org.hamcrest.Matchers.empty;
 import static org.hamcrest.Matchers.emptyIterable;
 import static org.hamcrest.Matchers.equalTo;
@@ -413,7 +412,7 @@
     }
 
     @SafeVarargs
-    protected final <T> void assertConnectionCount(
+    protected static <T> void assertConnectionCount(
             Iterable<T> addresses, Map.Entry<String, Integer>... addressAndConnCount) {
         @SuppressWarnings("unchecked")
         final Matcher<? super T>[] args = (Matcher<? super T>[]) Arrays.stream(addressAndConnCount)
@@ -422,7 +421,7 @@
                 .collect(Collectors.toList())
                 .toArray(new Matcher[] {});
         final Matcher<Iterable<? extends T>> iterableMatcher =
-                addressAndConnCount.length == 0 ? emptyIterable() : containsInAnyOrder(args);
+                addressAndConnCount.length == 0 ? emptyIterable() : contains(args);
         assertThat(addresses, iterableMatcher);
     }
 
@@ -437,7 +436,7 @@
                 .collect(Collectors.toList())
                 .toArray(new Matcher[] {});
         final Matcher<Iterable<? extends T>> iterableMatcher =
-                address.length == 0 ? emptyIterable() : containsInAnyOrder(args);
+                address.length == 0 ? emptyIterable() : contains(args);
         assertThat(addresses, iterableMatcher);
     }
 

```

Reply:
Huge thanks for putting so much effort into this. I reworked the concurrency as you suggested with more states but a single volatile variable that controls the lifetime. Please have a look. The test passes 10k iterations now.